dual_net.py

```diff
@@ -134,6 +134,40 @@ def should_log(i):
                     logger.add_summary(weight_summaries, global_step)
             self.save_weights()
 
+    def validate(self, tf_records, batch_size=128, logdir=None, init_from=None, num_steps=1000):
+        if logdir is None:
+            print("Need to run eval with a logdir")
+            return
+        if not logdir.endswith('_eval'):
+            logdir += "_eval"
+
+        with self.sess.graph.as_default():
+            input_tensors = preprocessing.get_input_tensors(
+                TRAIN_BATCH_SIZE, tf_records)
+            output_tensors = dual_net(input_tensors, TRAIN_BATCH_SIZE,
+                                      train_mode=False, **self.hparams)
+            train_tensors = train_ops(
+                input_tensors, output_tensors, **self.hparams)
+
+            # just run our cost tensors
+            validate_tensors = {k: train_tensors[k] for k in (
+                'policy_cost', 'value_cost', 'l2_cost',
+                'combined_cost')}
+            self.initialize_weights(init_from)
+            training_stats = StatisticsCollector()
+            logger = tf.summary.FileWriter(logdir, self.sess.graph)
+
+            for i in tqdm(range(num_steps)):
+                try:
+                    tensor_values = self.sess.run(validate_tensors)
+                except tf.errors.OutOfRangeError:
+                    break
+                training_stats.report(tensor_values)
+
+                if i % 10 == 9:
+                    accuracy_summaries = training_stats.collect()
```

Comment:
collect() should happen once at the very end, since you basically want to average all the eval() things into a single datapoint.

Reply:
Done.